src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -333,8 +336,36 @@ void ShenandoahConcurrentGC::entry_final_roots() {
   static const char* msg = "Pause Final Roots";
   ShenandoahPausePhase gc_phase(msg, ShenandoahPhaseTimings::final_roots);
   EventMark em("%s", msg);
-
+  ShenandoahHeap* const heap = ShenandoahHeap::heap();
   op_final_roots();
+
+  if (heap->mode()->is_generational()) {
```

Comment:
Thanks for asking these questions.  The existing ShenandoahHeap::rebuild_free_set() method can be called from here instead of this code.  I'm making this change.